File: bp-agent/src-rs/llm/router.rs
use std::collections::HashMap;
use std::sync::Arc;

use super::types::{CompletionRequest, LLMResponse, ProviderAdapter, ProviderError};

pub struct LLMRouter {
    default_provider: String,
    providers: HashMap<String, Arc<dyn ProviderAdapter>>,
}

impl LLMRouter {
    pub fn new(default_provider: &str) -> Self {
        Self {
            default_provider: default_provider.to_string(),
            providers: HashMap::new(),
        }
    }

    pub fn register_provider(&mut self, name: &str, adapter: Arc<dyn ProviderAdapter>) {
        self.providers.insert(name.to_string(), adapter);
    }

    pub fn complete(&self, request: CompletionRequest) -> Result<LLMResponse, ProviderError> {
        let provider = request
            .provider
            .clone()
            .unwrap_or_else(|| self.default_provider.clone());
        let adapter = self.providers.get(&provider).ok_or_else(|| {
            ProviderError::new("provider_missing", &format!("provider not registered: {}", provider), false)
        })?;
        adapter.complete(request)
    }
}

File: bp-agent/src/bp_agent/llm/router.py
"""Provider router."""

from __future__ import annotations

from typing import Protocol

from .types import CompletionRequest, LLMResponse, StreamChunk, StreamIterator


class ProviderAdapter(Protocol):
    def complete(self, request: CompletionRequest) -> LLMResponse:
        ...


class LLMRouter:
    def __init__(self, default_provider: str = "gemini"):
        self.default_provider = default_provider
        self._providers: dict[str, ProviderAdapter] = {}

    def register_provider(self, name: str, adapter: ProviderAdapter):
        self._providers[name] = adapter

    def complete(self, request: CompletionRequest) -> LLMResponse:
        provider = request.provider or self.default_provider
        if provider not in self._providers:
            raise ValueError(f"Provider not registered: {provider}")
        return self._providers[provider].complete(request)

    def complete_stream(self, request: CompletionRequest) -> StreamIterator:
        provider = request.provider or self.default_provider
        if provider not in self._providers:
            raise ValueError(f"Provider not registered: {provider}")
        adapter = self._providers[provider]
        if hasattr(adapter, "complete_stream"):
            return adapter.complete_stream(request)
        # Fallback: call complete() and yield a single chunk
        response = adapter.complete(request)
        return self._fallback_stream(response)

    @staticmethod
    def _fallback_stream(response: LLMResponse) -> StreamIterator:
        yield StreamChunk(delta=response.content, finish_reason="stop")
